bpo-21302: Add clock_nanosleep() implementation for time.sleep()

Livius90 opened the pull request:
- Add clock_nanosleep() function to time.sleep() in Unix operation systems.

Requested changes from @vstinner are done. It was runt-time tested in Oracle Linux 8.4 and Windows 10.
Some previous pull request history can be found here: https://github.com/python/cpython/pull/28077


<!-- issue-number: [bpo-21302](https://bugs.python.org/issue21302) -->
https://bugs.python.org/issue21302
<!-- /issue-number -->

Livius90 committed:
clock_nanosleep() implementation for time.sleep()

clock_nanosleep() is available in Linux which has POSIX 2001.12 or newer

vstinner commented on configure.ac:
```diff
@@ -4086,6 +4086,12 @@ AC_CHECK_FUNCS(clock_gettime, [], [
+    AC_CHECK_LIB(rt, clock_nanosleep, [
+        AC_DEFINE(HAVE_CLOCK_NANOSLEEP, 1)
+    ])
+])
```
Would you mind to move it after the clock_settime check?

vstinner commented on Modules/timemodule.c:
```diff
@@ -2066,14 +2070,24 @@ pysleep(_PyTime_t secs)
     }
     deadline = monotonic + secs;
+#ifdef HAVE_CLOCK_NANOSLEEP
+    if (_PyTime_AsTimespec(deadline, &timeout_abs) < 0)
```
Please add { ... } (PEP 7). Please also add them to the existing _PyTime_AsTimeval() call.

vstinner reviewed (commented):
Oh nice, it's much better!

IMO it's worth to mention that time.sleep() has a resolution of 1 nanosecond if the clock_nanosleep() function is available.

Do you know if clock_nanosleep() is available on other platforms than Linux? Is it available on macOS or FreeBSD for example?

Livius90 committed:
Follow PEP 7 rules in time.sleep()

PEP 7 rules appled for _PyTime_AsTimespec and _PyTime_AsTimeval error checking in pysleep().

Livius90 committed:
Adjust braces to PEP 7

Livius90 committed:
clock_nanosleep check placed after clock_settime

Livius90 committed:
Fix EINTR issue for clock_nanosleep()

In calling clock_nanosleep() EINTR is not stored in errno. Need to use return value of clock_nanosleep()/select() for checking it.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

Livius90 committed:
nanosleep() is available for Unix systems like OSX, FreeBSD

Livius90 committed:
Update news rst file

Unix operating systems eg: Linux, macOS, FreeBSD etc. time.sleep() has a resolution of nanoseconds with using clock_nanosleep() or nanosleep() function.

Livius90 committed:
replace select() with nanosleep()

In all Unix systems eg: macOS, FreeBSD, Linux etc, nanosleep() is available.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

Livius90 committed:
Fix test_eintr issue for macOS

Livius90 committed:
Remove news rst and regenerate

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

Livius90 committed:
Use waitable timer from Win32 for time.sleep() in Windows

Waitable timer is 100 nsec resolution. Now, seconds to nanosec conversion is limited in usec, soon will come the next developing part to improve it in next commit.

Livius90 committed:
Finalized waitable timer for time.sleep() in Windows

Waitable timer resolution is 100 nsec but it is limited to 1 usec by round ceiling, moreover sleep for lower then 1 milisec is not possible in Win32 API.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

Livius90 committed:
Fix macOS build issue

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

Livius90 committed:
Final optimization in Waitable timer object implementation

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

vstinner commented on Modules/_tkinter.c:
```diff
@@ -357,11 +357,10 @@ static int Tkinter_busywaitinterval = 20;
-    t.tv_usec = (milli%1000) * 1000;
-    select(0, (fd_set *)0, (fd_set *)0, (fd_set *)0, &t);
+    t.tv_nsec = (milli%1000) * 1000000;
+    nanosleep(&t, NULL);
```
Please revert this change and write a second PR for this change.

Livius90 replied:
Luckily nanosleep() is supported in many unix OS, see my latest comment.
https://github.com/python/cpython/pull/28111#issuecomment-914642746

Do you suggest to do it in a new PR for changes of [semaphore.c](https://github.com/python/cpython/pull/28111/files#diff-cf698f2997f9f11b2851514321974536c9a161585e85034230a9783c4a7bc98e), also?

vstinner commented on Misc/NEWS.d/next/Library/2021-09-04-11-46-05.bpo-21302.zTYR1A.rst:
```diff
@@ -0,0 +1 @@
+In Unix operating systems eg: Linux, macOS, FreeBSD etc. time.sleep() has a resolution of nanoseconds with using clock_nanosleep() or nanosleep() function. In Windows, time.sleep() has a resolution of microseconds with using waitable timer from Win32 API.
```
The Windosw code path is not modified, so you should not mention it in the *Change* log.

Livius90 replied:
Please look at my new Windows implementation of time.sleep(), too. I changed the sleep implementation in Windows, it is based on a Win32 waitable timer object instead of a dummy signal waiting and timeout.

vstinner reviewed (commented):
I would prefer to keep the select() code path since this one has been proved to work on all platforms (but Windows). If you want to use nanosleep(), I would prefer to add a configure check to check if the function is available.

Is it important for you to use nanosleep() on platforms where clock_nanosleep() is not available?

Maybe I'm wrong and nanosleep() is available "everywhere".

Livius90 committed:
revert to original

Livius90 committed:
remove news

Livius90 committed:
add clock_nanosleep() only

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

vstinner commented on Misc/NEWS.d/next/Library/2021-09-10-00-24-08.bpo-21302.XS_frc.rst:
```diff
@@ -0,0 +1 @@
+In Unix operating systems time.sleep() has a resolution of nanoseconds with using clock_nanosleep() function.
```
```suggestion
In Unix operating systems, :func:`time.sleep` now uses the ``clock_nanosleep()`` function,
if available, which allows to sleep for an interval specified with nanosecond precision.
```

vstinner commented on Modules/timemodule.c:
```diff
@@ -2053,7 +2053,11 @@ pysleep(_PyTime_t secs)
     _PyTime_t deadline, monotonic;
 #ifndef MS_WINDOWS
+#ifdef HAVE_CLOCK_NANOSLEEP
+    struct timespec timeout;
```
Usually, a timeout is relative. This one is special, it's absolute. Can you please rename it to "timeout_abs"?

Livius90 replied:
Done

vstinner commented on Modules/timemodule.c:
```diff
@@ -2066,23 +2070,44 @@ pysleep(_PyTime_t secs)
 
+#ifdef HAVE_CLOCK_NANOSLEEP
+        if (err != EINTR) {
+            errno = err;
```
Can you merge the two if/else code  paths? Only this line is different if HAVE_CLOCK_NANOSLEEP is defined.

Livius90 replied:
In case of using select(), errno must be compared in if statement. If clock_nanosleep() is used its (err) return value must be in the if statement comparing. If i am not following this, some unit tests will be failed because there are any magic behavier/differences in errno and the returned err of clock_nanosleep() and select(). I can not merge them better.

Livius90 replied:
I put errno value passing between separated Py_BEGIN_ALLOW_THREADS and Py_END_ALLOW_THREADS.

vstinner commented on Modules/timemodule.c:
```diff
@@ -2114,8 +2139,9 @@ pysleep(_PyTime_t secs)
             return -1;
+        }
 
         if (get_monotonic(&monotonic) < 0) {
```
Recomputing secs is useless if clock_nanosleep() is used no? It might be more efficient to avoid the get_monotonic() syscall.

Livius90 replied:
Yes, recomputing of monotonic time is totally useless if clock_nanosleep() is used.

Livius90 committed:
Update Misc/NEWS.d/next/Library/2021-09-10-00-24-08.bpo-21302.XS_frc.rst

suggested description

Co-authored-by: Victor Stinner <vstinner@python.org>

Livius90 committed:
timeout_abs renaming

Livius90 committed:
Simplified HAVE_CLOCK_NANOSLEEP macro blocks

Livius90 committed:
remake news

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

Livius90 committed:
Fix err checking

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

Livius90 committed:
revert changes

Can not simplify HAVE_CLOCK_NANOSLEEP blocks because do {} while loop need for EINTR error handling

Livius90 committed:
get_monotonic() and secs re-calculation are useless for clock_nanosleep()

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

Livius90 committed:
merge errno in HAVE_CLOCK_NANOSLEEP

Livius90 committed:
remake news rst

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

Livius90 committed:
Try to fix Address sanitizer test issue

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

Livius90 committed:
#2 Try to fix Address sanitizer test issue

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

Livius90 committed:
remake news rst

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

Livius90 committed:
re-make news rst

blurb-it[bot] committed:
📜🤖 Added by blurb_it.
